lib/SILOptimizer/Analysis/SimplifyInstruction.cpp

```diff
@@ -727,45 +746,556 @@ SILValue swift::simplifyInstruction(SILInstruction *I) {
+      if (newNotifyInst)
+        newNotifyInst(ebi);
+      continue;
+    }
+    case ValueOwnershipKind::None:
+      continue;
+    case ValueOwnershipKind::Unowned:
+      llvm_unreachable("Unowned object can never be consumed?!");
+    }
+    llvm_unreachable("Covered switch isn't covered");
+  }
+}
+
+static ValueOwnershipKind findCompatibleNonNoneOwnershipKind(OperandOwnershipKindMap map) {
+  assert(map.canAcceptKind(ValueOwnershipKind::None) && "We should always have maps that can accept none");
+
+  // Ok, we can accept None. Lets see if we can't accept any other type of
+  // ownership.
+  if (map.canAcceptKind(ValueOwnershipKind::Guaranteed)) {
+    return ValueOwnershipKind::Guaranteed;
+  }
+
+  if (map.canAcceptKind(ValueOwnershipKind::Owned)) {
+    return ValueOwnershipKind::Owned;
+  }
+
+  if (map.canAcceptKind(ValueOwnershipKind::Unowned)) {
+    return ValueOwnershipKind::Unowned;
+  }
+
+  return ValueOwnershipKind::None;
+}
+
+SILBasicBlock::iterator OwnershipRAUWUtility::handleReplaceOfNone() {
+  auto newKind = newValue.getOwnershipKind();
+  SmallVector<Operand *, 8> nonConsumingUses;
+  SmallVector<Operand *, 8> consumingUses;
+  SmallVector<Operand *, 8> convertToGuaranteed;
+  SmallVector<Operand *, 8> convertToOwned;
+  SmallVector<Operand *, 8> convertToUnowned;
+  SmallVector<Operand *, 8> onlyAcceptNone;
+
+  for (auto *use : oldValue->getUses()) {
+    if (use->isTypeDependent())
+      continue;
+    auto map = use->getOwnershipKindMap();
+    // If we can accept our new kind...
+    if (map.canAcceptKind(newKind)) {
+      // It isn't a liveness use, we can just lifetime extend.
+      if (map.getLifetimeConstraint(newKind) == UseLifetimeConstraint::MustBeLive) {
+        nonConsumingUses.push_back(use);
+      } else {
+        // If it is a consuming use, we may need to insert compensating
+        // instructions here along with lifetime extension.
+        consumingUses.push_back(use);
+      }
+      continue;
+    }
+
+    // Ok, we need to perform some sort of conversion for this value. Lets look
+    // at the map to find out what it can accept. First see if we can only
+    // accept .none. In that case, we may need to deal with forwarding
+    // instructions.
+    switch (findCompatibleNonNoneOwnershipKind(map)) {
+    case ValueOwnershipKind::None:
+      onlyAcceptNone.push_back(use);
+      continue;
+    case ValueOwnershipKind::Owned:
+      convertToOwned.push_back(use);
+      continue;
+    case ValueOwnershipKind::Guaranteed:
+      convertToGuaranteed.push_back(use);
+      continue;
+    case ValueOwnershipKind::Unowned:
+      convertToUnowned.push_back(use);
+      continue;
+    }
+  }
+
+  // Ok, we have categorized our uses. First handle the conversion uses.
+  assert(onlyAcceptNone.empty() && convertToOwned.empty() && convertToGuaranteed.empty() && convertToUnowned.empty());
+
+  // Then handle the non-conversion uses.
+  switch (newKind) {
```

Comment:
I don't know why you need to separately switch over the operand ownership constraint above and switch over the new kind here. The new operand kind should be determined directly by joining the new value kind with the operand constraint. Then you switch over OwnershipKindPair reusing the same code in `::perform` but without any of the ".none" cases.